File: Java/src/main/java/com/gildedrose/Item.java
package com.gildedrose;

public class Item {

    public static final int MAX_QUALITY_VALUE = 50;
    public String name;

    public int sellIn;

    public int quality;

    public Item(String name, int sellIn, int quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }

    @Override
    public String toString() {
        return this.name + ", " + this.sellIn + ", " + this.quality;
    }

    public void improveQuality() {
        if (quality < MAX_QUALITY_VALUE) {
            quality = quality + 1;
        }
    }

    public String getName() {
        return name;
    }


    public int getSellIn() {
        return sellIn;
    }

    public void setSellIn(int sellIn) {
        this.sellIn = sellIn;
    }


    public void setQuality(int quality) {
        this.quality = quality;
    }

    public void degradeQuality() {
        quality -= quality > 0 ? 1 : 0;
        quality -= sellIn < 0 ? 1 : 0;
    }

    public int getQuatity() {
        return quality;
    }
}

File: Kotlin/src/main/kotlin/com/gildedrose/Item.kt
package com.gildedrose

data class Item(var name: String, var sellIn: Int, var quality: Int)


const val MAX_QUALITY = 50
const val MIN_QUALITY = 0

const val MIN_SELLIN = 0

inline fun Item.degradeQuality() {
    if (quality > MIN_QUALITY) quality -= 1
}


inline fun Item.increaseQuality() {
    if (quality < MAX_QUALITY) quality += 1
}


inline fun Item.getOlder() {
    sellIn -= 1
}

inline fun Item.isExpired(): Boolean = sellIn < MIN_SELLIN


inline fun Item.process() {
    ProcessorFactory.createProcessor(this).process(this)
}